### PR TITLE
Display message when too many annotations to show in Notebook

### PR DESCRIPTION
**Pull request by lyzadanger**
This draft PR starts the work of showing a user-friendly message if there are too many annotations in a group to show within the Notebook at present (current max: 5000).

Initial screenshots. _Note_: Wording is totally preliminary. More on that shortly. Also note that I've dialed down the `maxResults` to 5 here to be able to show the error, but it would normally be 5000.

Wider screens:

image

Narrower screens:

image

### Implementation Notes

* Makes `LoadAnnotationService#load` take an optional `onError` handler for when the `SearchClient` emits an error. This allows for alternate error handling than `console.error`
* Updates `search-client` to emit a `resultCount` before an `error` on `maxResults` to allow handling code to know how many annotations it was _trying_ to retrieve.
* Adds a generic `Panel` component that is also used by `SidebarPanel`
* Updates the `NotebookView` to render a `Panel` if there are too many annotations to show.

### Next Steps

* [x] Get product owner input on how to word this and stylistic approval. I can contribute to wordsmithing but need some guidance about level of levity allowed, tone, whatnot.
* [x] Get technical feedback and general acceptance on approach
* [x] Write tests and clean up

When it's ready and lands it:

Fixes https://github.com/hypothesis/client/issues/3124

### CODE REVIEW

File: src/sidebar/components/NotebookView.js [hunk 1]
```diff
@@ -37,11 +38,20 @@ function NotebookView({ loadAnnotationsService }) {
...
   const [paginationPage, setPaginationPage] = useState(1);
 
+  const [hasLoadError, setHasLoadError] = useState(false);
+
+  // Load all annotations in the group, unless there are more than 5000
+  // of them: this is a performance safety valve.
+  const maxResults = 5000;
+
+  const onLoadError = error => {
+    if (error.message?.includes('Results size exceeds')) {
```
Comment (robertknight) [inline src/sidebar/components/NotebookView.js:48]: I would suggest defining a specific `Error` subclass for this error. The risk with testing the `message` string is that someone changes it (as `message` is not in general intended to be machine-readable) and doesn't realize that it is important.

For example:

```js
class ResultSizeError extends Error {
  constructor(limit) {
    super(`Results size exceeds ${limit}`);
  }
}
```

Then:

```js
if (error instanceof ResultSizeError) {
  ...
}
```

A variation on the above is to define a more general error subclass for use in some context and then add a property (eg. `type`) with a value that is specifically intended to be machine-readable.

As a side benefit, this provides somewhere to hang additional error properties if you need them.

File: src/sidebar/components/NotebookView.js [hunk 2]
```diff
@@ -93,6 +104,16 @@ function NotebookView({ loadAnnotationsService }) {
         />
       </div>
       <div className="NotebookView__items">
+        {hasLoadError && (
```
Comment (robertknight) [inline src/sidebar/components/NotebookView.js:107]: The state name `hasLoadError` is quite generic and suggests that _some_ kind of error occurred, but the message is specific to a particular error. Some possible solutions:

1. Make `loadError` an `Error|null` which is set to the actual `Error` if one occurred and then test the type when rendering
2. Make `loadError` an enum or null (eg. `'too-many-results'|'some-other-error'|null`) which specifies the type of error that occurred

File: src/sidebar/components/Panel.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Render a simple, consistent "panel"-like interface
+ *
+ * @param {PanelProps} props
+ */
+export default function Panel({ title, children }) {
+  return (
+    <div className="Panel">
+      <div className="Panel__header">
+        <div className="Panel__title">{title}</div>
```
Comment (robertknight) [inline src/sidebar/components/Panel.js:16]: The corresponding element in `SidebarPanel` uses a semantic `h2` for the heading. I could imagine that in future we might make other small visual or a11y changes to `SidebarPanel` and forget to incorporate them in `Panel`.

Given that the visual structure of the `Panel` is the same as `SidebarPanel`, except for the lack of a close button, perhaps it would make sense to refactor this so that:

- Panel is a simple/pure component that renders the panel UI with a set of props
- SidebarPanel wraps panel to add the sidebar-specific integration

In other words, this would be conceptually similar to say `Menu` vs `UserMenu`.

Thoughts?

**Review by robertknight — COMMENTED** [review]
I gave some feedback on the details. Two main items:

- Testing the `message` of an error is a refactoring hazard IMO because this value is not, in general, intended to be machine readable and so it is easy for someone to change the message without realizing that downstream logic might depend on it
- I spotted what looked like an unintentional semantic difference in the structure of the Panel vs. SidebarPanel HTML. This made me think whether perhaps `SidebarPanel` should wrap/compose Panel to add the sidebar-specific logic rather than being an entirely unrelated component.

**Comment by klemay** [comment]
Feedback re: proposed language - it would be nice if we gave people the option to be emailed when the ability to support >5k annotations is in place. Something like:

Too many results to show

This preview of the Notebook works in courses with up to 5,000 results (there are 6,000 to show in this course). Support for more annotations is coming soon. [Contact us](http://support@hypothes.is?subject=Hypothesis%20Notebook&body=Please%20notify%20me%20when%20the%20Hypothesis%20Notebook%20is%20updated%20to%20support%20courses%20with%20more%20than%205000%20annotations) to be notified when the Notebook is updated accordingly.

The support team has agreed to tag those incoming emails so I can easily find them & notify folks when the time is right. 

I do think my proposed "...when the Notebook is updated accordingly" is a bit awkward and could use further wordsmithing.

**Comment by lyzadanger** [comment]
@klemay At this time, the Notebook doesn't "know" whether it's operating within an LMS context. Given that we're trying to work in a context-agnostic way, I had been avoiding the use of LMS-specific terminology like "course".* "Course" is also slightly misleading as the user might be looking at activity within a single section group (subset of _course_, in that sense). 

Here's a riff:

> This preview of the Notebook can show up to 5,000 results (there are 6,000 to show here). [Contact us](http://support@hypothes.is/?subject=Hypothesis%20Notebook&body=Please%20notify%20me%20when%20the%20Hypothesis%20Notebook%20is%20updated%20to%20support%20courses%20with%20more%20than%205000%20annotations) if you’d like to be notified when support for more annotations is available.

\* I don't want to over-force this vocabulary generalization if it becomes more useful to use LMS-specific language in certain contexts as we go, but the Notebook doesn't have at present any contextual awareness.

**Comment by klemay** [comment]
@lyzadanger point taken, and I think your proposed edit would work perfectly!

**Comment by lyzadanger** [comment]
@robertknight Since first technical review:

* `ResultSizeError` Error class used instead of lame string matching (embarrrrasssssed)
* `Panel` refactored; `SidebarPanel` uses `Panel` now, no more divergent styling/semantics
*  Approved wording used
* Tests added

File: src/sidebar/components/NotebookView.js [hunk 4]
```diff
@@ -115,6 +128,20 @@ function NotebookView({ loadAnnotationsService }) {
         />
       </div>
       <div className="NotebookView__items">
+        {hasTooManyAnnotationsError && (
+          <div className="NotebookView__messages">
+            <Panel title="Too many results to show">
+              This preview of the Notebook can show{' '}
+              <strong>up to {maxResults} results</strong> at a time (there are{' '}
+              {resultCount} to show here).{' '}
+              <a href="mailto:support@hypothes.is?subject=Hypothesis%20Notebook&body=Please%20notify%20me%20when%20the%20Hypothesis%20Notebook%20is%20updated%20to%20support%20more%20than%205000%20annotations">
```
Comment (lyzadanger) [inline src/sidebar/components/NotebookView.js:137]: Does this need better encoding?

File: src/sidebar/components/test/NotebookView-test.js [hunk 5]
```diff
@@ -3,6 +3,7 @@ import { act } from 'preact/test-utils';
 
 import mockImportedComponents from '../../../test-util/mock-imported-components';
 
+import { ResultSizeError } from '../../search-client';
```
Comment (lyzadanger) [inline src/sidebar/components/test/NotebookView-test.js:6]: This does not make me happy. Any other ideas?

File: src/sidebar/components/test/NotebookView-test.js [hunk 6]
```diff
@@ -91,6 +94,20 @@ describe('NotebookView', () => {
     assert.notCalled(fakeLoadAnnotationsService.load);
   });
 
+  it('shows a message if too many annotations to load', () => {
+    // Simulate the loading service emitting an error indicating
+    // too many annotations to load
+    fakeLoadAnnotationsService.load.callsFake(options => {
+      options.onError(new ResultSizeError(5000));
```
Comment (lyzadanger) [inline src/sidebar/components/test/NotebookView-test.js:101]: `onError` needs to be called with a `ResultSizeError` to trigger `NotebookView`'s messaging. Ergo the dependency on `ResultSizeError`. But...hmmm.

**Review by robertknight — APPROVED** [review]
I think you can simplify the `onClose`/`withCloseButton` props into a single prop (just don't show a button if no `onClose` handler is passed). I also found an apparently unused `SidebarPanel__container` class. Otherwise LGTM.

File: src/sidebar/components/SidebarPanel.js [hunk 7]
```diff
@@ -58,25 +54,21 @@ export default function SidebarPanel({
...
-              <SvgIcon name={icon} title={title} />
-            </div>
-          )}
-          <h2 className="SidebarPanel__title u-stretch">{title}</h2>
-          <div>
-            <Button icon="cancel" buttonText="Close" onClick={closePanel} />
-          </div>
-        </div>
-        <div className="SidebarPanel__content">{children}</div>
+      <div className="SidebarPanel__container" ref={panelElement}>
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/SidebarPanel.js:63]: Indeed it can go!

File: src/sidebar/search-client.js [hunk 8]
```diff
@@ -116,6 +130,12 @@ export default class SearchClient extends TinyEmitter {
         return;
       }
 
+      if (this._resultCount === null) {
+        // Emit the result count (total) on first encountering it
+        this._resultCount = results.total;
+        this.emit('resultCount', this._resultCount);
+      }
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/search-client.js:137]: Exactly correct. It's to ensure that value gets set on the store so components have access to how many results it was _trying_ to fetch.
